libdnf/transaction/transaction.cpp

```diff
@@ -206,6 +205,31 @@ void Transaction::fill_transaction_packages(
 }
 
 
+void Transaction::fill_transaction_groups(
+    const std::vector<libdnf::base::TransactionGroup> & transaction_groups,
+    const std::set<std::string> & installed_names) {
+    for (auto & tsgrp : transaction_groups) {
+        auto & new_grp = new_comps_group();
+        auto group = tsgrp.get_group();
+        new_grp.set_name(group.get_name());
+        new_grp.set_action(tsgrp.get_action());
+        new_grp.set_reason(tsgrp.get_reason());
+        new_grp.set_translated_name(group.get_translated_name());
+
+        libdnf::comps::PackageType package_types{0};
+        for (const auto & group_package : group.get_packages()) {
+            auto & new_grp_pkg = new_grp.new_package();
+            auto name = group_package.get_name();
+            new_grp_pkg.set_name(name);
+            new_grp_pkg.set_package_type(group_package.get_type());
+            new_grp_pkg.set_installed(installed_names.contains(name) ? true : false);
+            package_types |= group_package.get_type();
+        }
```

Comment:
May be it will sound strange but what about installation of a group without optional packages but I believe that the code will also mark all installed packages from optional group as group installed (the same problem can be achieved using multiple approaches). May be it would be nice to track more precisely what was installed by particular group. Or to use lass precise model for storing data. Somehow I feel that it is something in the middle. But it can be resolved by TODO.